internal/controller/controllers/agentserviceconfig_controller_storage_validation_test.go

```diff
@@ -0,0 +1,431 @@
+	clnt "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("Agent service config controller storage validation", func() {
+	var (
+		ctx        context.Context
+		cluster    *testing.FakeCluster
+		client     clnt.Client
+		subject    *aiv1beta1.AgentServiceConfig
+		key        clnt.ObjectKey
+		request    ctrl.Request
+		reconciler *AgentServiceConfigReconciler
+	)
+
+	BeforeEach(func() {
+		// Create a context for the test:
+		ctx = context.Background()
+
+		// Create the fake cluster:
+		cluster = testing.NewFakeCluster().
+			Logger(logger).
+			Build()
+
+		// Create the client:
+		client = cluster.Client()
+
+		// Create the default object to be reconciled, with a storage configuration that
+		// passes all validations. Specific tests will later adapt these settings as needed.
+		subject = &aiv1beta1.AgentServiceConfig{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "agent",
+			},
+			Spec: aiv1beta1.AgentServiceConfigSpec{
+				DatabaseStorage: corev1.PersistentVolumeClaimSpec{
+					AccessModes: []corev1.PersistentVolumeAccessMode{
+						corev1.ReadWriteOnce,
+					},
+					Resources: corev1.ResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceStorage: resource.MustParse("1Ti"),
+						},
+					},
+				},
+				FileSystemStorage: corev1.PersistentVolumeClaimSpec{
+					AccessModes: []corev1.PersistentVolumeAccessMode{
+						corev1.ReadWriteOnce,
+					},
+					Resources: corev1.ResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceStorage: resource.MustParse("1Ti"),
+						},
+					},
+				},
+				ImageStorage: &corev1.PersistentVolumeClaimSpec{
+					AccessModes: []corev1.PersistentVolumeAccessMode{
+						corev1.ReadWriteOnce,
+					},
+					Resources: corev1.ResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceStorage: resource.MustParse("1Ti"),
+						},
+					},
+				},
+			},
+		}
+		key = clnt.ObjectKeyFromObject(subject)
+		request = ctrl.Request{
+			NamespacedName: types.NamespacedName{
+				Name: subject.Name,
+			},
+		}
+
+		// Create the reconciler:
+		reconciler = &AgentServiceConfigReconciler{
+			AgentServiceConfigReconcileContext: AgentServiceConfigReconcileContext{
+				Client:   client,
+				Scheme:   cluster.Scheme(),
+				Log:      logrusLogger,
+				Recorder: cluster.Recorder(),
+			},
+			Namespace: "assisted-installer",
+		}
+	})
+
+	AfterEach(func() {
+		cluster.Stop()
+	})
+
+	When("Objects haven't been created", func() {
+		It("Rejects small database storage", func() {
+			subject.Spec.DatabaseStorage.Resources.Requests = corev1.ResourceList{
+				corev1.ResourceStorage: resource.MustParse("1Mi"),
+			}
+			err := client.Create(ctx, subject)
+			Expect(err).ToNot(HaveOccurred())
+			Eventually(func(g Gomega) {
+				_, err := reconciler.Reconcile(ctx, request)
+				g.Expect(err).ToNot(HaveOccurred())
+				err = client.Get(ctx, key, subject)
+				g.Expect(err).ToNot(HaveOccurred())
```

Comment:
Any reason to use the `g` instance here rather than just the global `Expect` function?

Reply:
Using the `g.Expect(...)` here is needed because using just `Expect(...)` here will indicate to the `gomega` library that the the test has failed completely. What we want to indicate here is that the current iteration inside `Eventually(...)` failed, so that the library will try again.